src/libraries/System.Net.Http/src/System/Net/Http/Metrics/HttpMetricsEnrichmentContext.cs

```diff
@@ -0,0 +1,171 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Net.Http.Metrics
+{
+    /// <summary>
+    /// Provides functionality for enriching request metrics `http-client-request-duration` and `http-client-failed-requests`.
+    /// </summary>
+    /// <remarks>
+    /// Enrichment is done on per-request basis by callbacks registered with <see cref="AddCallback(HttpRequestMessage, Action{HttpMetricsEnrichmentContext})"/>.
+    /// The callbacks are responsible for adding custom tags via <see cref="AddCustomTag(string, object?)"/> for which they can use the request, response and error
+    /// information exposed on the <see cref="HttpMetricsEnrichmentContext"/> instance.
+    ///
+    /// > [!IMPORTANT]
+    /// > The <see cref="HttpMetricsEnrichmentContext"/> intance must not be used from outside of the enrichment callbacks.
+    /// </remarks>
+    public sealed class HttpMetricsEnrichmentContext
+    {
+        private static readonly HttpRequestOptionsKey<HttpMetricsEnrichmentContext> s_optionsKeyForContext = new("HttpMetricsEnrichmentContext");
+        private static readonly ConcurrentQueue<HttpMetricsEnrichmentContext> s_pool = new();
+        private static int s_poolItemCount;
+        private const int PoolCapacity = 1024;
+
+        private readonly List<Action<HttpMetricsEnrichmentContext>> _callbacks = new();
```

Comment:
This is based on R9's [common pooling logic](https://github.com/search?q=repo%3Adotnet%2Fextensions+PoolFactory&type=code), using a bounded `ObjectPool`. The trick is that we pool the callback list together with the `HttpMetricsEnrichmentContext` in a thread-safe manner. The resulting code seems to simpler to me than the combination of pooling a context in `[ThreadStatic]` & implementing a special type to store the delegate list while eliminating more per-request allocations in the long run.

## Benchmarks

### 1. Sending a single GET request

#### Without pooling

```
|                   Method |     Mean |    Error |   StdDev | Ratio | Allocated | Alloc Ratio |
|------------------------- |---------:|---------:|---------:|------:|----------:|------------:|
|                NoMetrics | 97.20 ms | 0.301 ms | 0.267 ms |  1.00 |   4.14 KB |        1.00 |
|              WithMetrics | 97.25 ms | 0.141 ms | 0.132 ms |  1.00 |   4.42 KB |        1.07 |
| WithMetricsAndEnrichment | 97.20 ms | 0.190 ms | 0.169 ms |  1.00 |   5.27 KB |        1.27 |
```

#### With pooling

```
|                   Method |     Mean |    Error |   StdDev | Ratio | Allocated | Alloc Ratio |
|------------------------- |---------:|---------:|---------:|------:|----------:|------------:|
|                NoMetrics | 97.19 ms | 0.232 ms | 0.206 ms |  1.00 |   4.07 KB |        1.00 |
|              WithMetrics | 97.42 ms | 0.469 ms | 0.438 ms |  1.00 |   4.36 KB |        1.07 |
| WithMetricsAndEnrichment | 97.19 ms | 0.174 ms | 0.163 ms |  1.00 |    4.8 KB |        1.18 |
```

### 2. Sending 20 GET requests in parallel

#### Without pooling

```
|                   Method | ParallelRequestCount |     Mean |   Error |  StdDev | Ratio | Allocated | Alloc Ratio |
|------------------------- |--------------------- |---------:|--------:|--------:|------:|----------:|------------:|
|                NoMetrics |                   20 | 101.7 ms | 0.79 ms | 0.66 ms |  1.00 |   79.9 KB |        1.00 |
|              WithMetrics |                   20 | 101.9 ms | 0.66 ms | 0.59 ms |  1.00 |  82.51 KB |        1.03 |
| WithMetricsAndEnrichment |                   20 | 102.2 ms | 0.72 ms | 0.64 ms |  1.00 |  100.8 KB |        1.26 |
```

#### With pooling


```
|                   Method | ParallelRequestCount |     Mean |   Error |  StdDev | Ratio | Allocated | Alloc Ratio |
|------------------------- |--------------------- |---------:|--------:|--------:|------:|----------:|------------:|
|                NoMetrics |                   20 | 101.9 ms | 0.66 ms | 0.55 ms |  1.00 |  79.92 KB |        1.00 |
|              WithMetrics |                   20 | 102.3 ms | 0.54 ms | 0.50 ms |  1.00 |  82.92 KB |        1.04 |
| WithMetricsAndEnrichment |                   20 | 101.5 ms | 0.73 ms | 0.68 ms |  1.00 |   91.4 KB |        1.14 |
```

Looks like this reduced allocations measurably without introducing contention. Further reduction is possible by optimizing `HttpRequestOptions` storage.

Reply:
Couple questions:
- Can I see the benchmark code that produces this output somewhere?
- Do you know where the allocations are coming from in the code?
- For latency measurements, is this how the networking team usually does it? I'm guessing all the numbers we see for latency are measuring the noise in network transmission which is orders of magnitude larger than the CPU effects. Doing a test over loopback with high parallelism and low message size seems like it could put tighter bounds on CPU overhead and potential cross-thread contention.

Reply:
Oops, I forgot to link the benchmarks: https://gist.github.com/antonfirsov/6bcd1e2d099eef52a077dd0fab862bfb

> Do you know where the allocations are coming from in the code?

Given that the enrichment-related data structures are pooled now, I *think* the difference between `WithMetrics` and `WithMetricsAndEnrichment` comes mostly from `HttpRequestOptions`, which is not being allocated when there are no enrichment callbacks registered.

What's more problematic is that I would expect the difference between `NoMetrics` and `WithMetrics` to be negligible which is not the case. I assume the ValueTask state machine in [ `SendAsyncWithMetrics`](https://github.com/dotnet/runtime/blob/6d4cd18fe03951f7fffd3a869bd7a69ac80f2509/src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs#L53-L74) should be completely allocation-free right? That method should not allocate then. Maybe it's the instruments & MeterListener?

> latency are measuring the noise in network transmission

Good point, I will change my benchmarks to use a loopback server instead.

Reply:
@MihaZupan do you see anything suspicious?

Reply:
Ok, I will play with `PoolingAsyncValueTaskMethodBuilder` and improve my benchmarks then.

Reply:
I think it's okay as-is too. We've historically been pretty selective with where we use the pooling builder.
That is, I don't see why this particular method would be any more/less suitable for it than some of our existing ones inside `SocketsHttpHandler` where we don't currently use it.

Reply:
I have changed my benchmarks dummying out all network traffic by using `ConnectCallback` and a stream producing a simple HTTP/1.1 response, see the [updated gist](https://gist.github.com/antonfirsov/6bcd1e2d099eef52a077dd0fab862bfb) for the code.

The differences between the parallel request benchmark times do not seem to be significant, or within the margin of error.

### Without caching

```
|                   Method | ParallelRequestCount |     Mean |    Error |   StdDev | Ratio | RatioSD |    Gen0 |   Gen1 | Allocated | Alloc Ratio |
|------------------------- |--------------------- |---------:|---------:|---------:|------:|--------:|--------:|-------:|----------:|------------:|
|                NoMetrics |                   20 | 39.86 us | 0.760 us | 1.205 us |  1.00 |    0.00 | 19.6533 | 3.1738 | 164.88 KB |        1.00 |
|              WithMetrics |                   20 | 44.20 us | 0.840 us | 1.947 us |  1.12 |    0.06 | 20.0195 | 3.1738 | 170.85 KB |        1.04 |
| WithMetricsAndEnrichment |                   20 | 50.29 us | 0.626 us | 0.696 us |  1.27 |    0.03 | 23.6816 | 4.3945 |  200.8 KB |        1.22 |
```

### With caching

```
|                   Method | ParallelRequestCount |     Mean |    Error |   StdDev |   Median | Ratio | RatioSD |    Gen0 |   Gen1 | Allocated | Alloc Ratio |
|------------------------- |--------------------- |---------:|---------:|---------:|---------:|------:|--------:|--------:|-------:|----------:|------------:|
|                NoMetrics |                   20 | 40.75 us | 0.807 us | 1.980 us | 40.31 us |  1.00 |    0.00 | 19.6533 | 3.1738 | 165.74 KB |        1.00 |
|              WithMetrics |                   20 | 45.10 us | 1.133 us | 3.214 us | 44.02 us |  1.12 |    0.10 | 20.0195 | 3.1738 | 170.81 KB |        1.03 |
| WithMetricsAndEnrichment |                   20 | 50.89 us | 0.852 us | 1.222 us | 50.66 us |  1.22 |    0.05 | 22.7051 | 3.6621 | 191.66 KB |        1.16 |
```